### PR TITLE
fix job update labels issue

### LINKED ISSUE #31009: Kubectl Get jobs with label selector returns blank
**Kubernetes version** (use `kubectl version`):
1.3.4

**What happened**:
Performing a `kubectl get jobs -l <any label selector>` returns no jobs despite their existence.

**What you expected to happen**:
List of jobs matching the label selector

**How to reproduce it** (as minimally and precisely as possible):
Create a job with a label. Call `kubectl get jobs -l <selector>`

**Anything else do we need to know**:
I am not sure if this is similar to #22517 or not

### PR DESCRIPTION
**Pull request by adohe-zz**
fix #31009 

@soltysh ptal. I want to make sure this is correct way to do this.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32033)

### CODE REVIEW

**Comment by floreks** [comment]
I've tested it and I can confirm that it works.

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/apis/batch/v1/defaults.go, line 44 \[r1\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/32033#-KQtA9R7_bJcbQbG6Koh:-KQtA9R7_bJcbQbG6Koi:bytmd07) ([raw file](https://github.com/kubernetes/kubernetes/blob/c828bf8ba9dddf6f40033103431916635ada64ca/pkg/apis/batch/v1/defaults.go#L44)):*
> ```Go
> 	// When manualSelector is true, set metadata.Labels according to
> 	// obj.Spec.Template.Labels
> 	labels := obj.Spec.Template.Labels
> ```

I see that this is similar to what's done in petset defaults. Shouldn't label selector be also set same as [here](https://github.com/kubernetes/kubernetes/blob/7457166290b636d08e01c45baf8e7ecc7744755b/pkg/apis/apps/v1alpha1/defaults.go#L33)?

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/32033#-:-KQtAvwoCGiR1ef0L1h0:b4jhp4c)*

File: pkg/apis/batch/v1/defaults.go [hunk 1]
```diff
@@ -39,4 +39,12 @@ func SetDefaults_Job(obj *Job) {
...
 		*obj.Spec.Parallelism = 1
 	}
+	// When manualSelector is true, set metadata.Labels according to
+	// obj.Spec.Template.Labels
+	labels := obj.Spec.Template.Labels
+	if labels != nil {
+		if len(obj.Labels) == 0 {
+			obj.Labels = labels
+		}
+	}
```
Comment (soltysh) [inline pkg/apis/batch/v1/defaults.go]: I'm not sure we want to copy pod's template labels to the job, if none are found. I need to dig into what's the root cause of those missing labels. But at this point this is not what I'd expect it to be.

**Comment by soltysh** [comment]
@AdoHe thanks for digging, but see https://github.com/kubernetes/kubernetes/issues/31009#issuecomment-244763582. I'll close this PR in favor of digging into the root cause.

**Comment by soltysh** [comment]
@AdoHe apologies apparently I was wrong, since almost forever (#6616) we are defaulting object's labels with pod template's one if none are specified. With that in mind, this PR LGTM. I'll prepare a followup for `batch/v2alpha1` and I'll think what about ScheduledJobs.

File: pkg/apis/batch/v1/defaults.go [hunk 2]
```diff
@@ -39,4 +39,12 @@ func SetDefaults_Job(obj *Job) {
 		obj.Spec.Parallelism = new(int32)
 		*obj.Spec.Parallelism = 1
 	}
+	// When manualSelector is true, set metadata.Labels according to
```
Comment (soltysh) [inline pkg/apis/batch/v1/defaults.go]: If you get a chance before the bot merges it, fix the comment. In `batch/v1` there's no `manualSelector`, only automatic one. Generally no comment here is needed.

File: pkg/apis/batch/v1/defaults.go [hunk 3]
```diff
@@ -39,4 +39,12 @@ func SetDefaults_Job(obj *Job) {
...
+	// obj.Spec.Template.Labels
+	labels := obj.Spec.Template.Labels
+	if labels != nil {
```
Comment (soltysh) [inline pkg/apis/batch/v1/defaults.go]: Squash the if:
```
if labels != nil && len(obj.Labels) == 0 {
	obj.Labels = labels
}
```

**Comment by soltysh** [comment]
Closing in favor of #32110.

**Pull request closed without merging**